### DISCUSSION TITLE
strongly typed C++ adaptors for parser grammars

### DISCUSSION THREAD

**Discussion by ikrima** [tree-sitter/tree-sitter#2982] [Libraries & tools]
# TLDR
- Dev quality of life C++ bindings out of a custom grammar for walking/traversing nodes
- Don't think this really belongs in tree-sitter main but sharing for fun/maybe others have similar uses

---

I love the bare C, dependency free nature of tree-sitter. But overtime, found traversing the parse tree being a chore especially since tree-sitter is used as the primary layer for parsing/AST manipulation in a custom compiler (i.e. not just syntax highlighting)

## Needs
- traversing the AST for semantic analysis
- lowering to mlir/llvm
- generating AST transformers, UI debug, etc

## Result 
So hacked together a little proof of concept generator from the node-types.js + made some C++ adapters
_caveat: it's hackishly ghetto so ignore the mess of code, especially the js. not a js dev_

From a day of use, it's been a net positive

I liked the mlir architecture so applied the same model (ie the strong typed Tree-Sitter nodes are just adapters of the core TSNode so they can just be typecast back & forth, modulo special compiler settings of struct layout/packing)

### Sample Use

```cpp
    CSTree_o cstree = s_DSLSubsys->parseCSTree(srcbuf);
    Opdef_cst cstnde(cstree.root());
    for (Oprm_cst& oprm : cstnde.oprmblkNde().childNodes()) {
      prmblk.push_back(mbldr.prmbndAttr(as_strref(oprm.nameNde().src(srcbuf)),
        mlirGenEnum<EOprmFlag>(oprm.oprmetaNde().prmflagsNde()), mlirGenType(oprm.typeNde())));
    }
    OpdefOp opdefop = mbldr.opdefOp(as_strref(cstnde.evalFnNde().src(srcbuf)), prmblk);
```

### Code
hacked generator + generated code for a custom c like grammar
https://gist.github.com/ikrima/3345769249993cafd7851390966089cb
